README.md
# MaciASL-Xc233
A simplified version of MaciASL that based on acidanthera's repo.

## Changelog
#### v1.0.0

- Remove useless menu
- Remove Log/Summary/Print function
- Simplified Preferences Windows
- Remove SSDT Generater
- Only Rehabman laptop and VoodooI2C patches
- Remove Update button and aisle version brief introduction


Changelog.md
MaciASL-Xc233 Changelog
=================
#### v1.0.0

- Remove useless menu
- Remove Log/Summary/Print
- Simplified Preferences Windows
- Only Rehabman laptop and VoodooI2C patches
- Remove Update button and aisle version brief introduction






MaciASL/DocumentController.m
//
//  DocumentController.m
//  MaciASL
//
//  Created by PHPdev32 on 10/14/13.
//  Licensed under GPLv3, full text at http://www.gnu.org/licenses/gpl-3.0.txt
//

#import "DocumentController.h"
#import "Document.h"
#import "iASL.h"

@implementation DocumentController {
    @private
    __unsafe_unretained IBOutlet NSWindow *_tableView;
}

#pragma mark NSDocumentController
-(id)makeDocumentWithContentsOfURL:(NSURL *)url ofType:(NSString *)typeName error:(NSError *__autoreleasing *)outError {
    id document;
    NSString *table;
    if ([typeName isEqualToString:kUTTypeTableset]) {
        if ((document = [self openTablesetTableWithContentsOfURL:url]))
            [self noteNewRecentDocumentURL:url];
    }
    else if ([typeName isEqualToString:kUTTypeIOJones]) {
        if ((document = [self openIOJonesTableWithContentsOfURL:url]))
            [self noteNewRecentDocumentURL:url];
    }
    else if ((table = [iASL isInjected:url])) {
        iASLDecompilationResult *decompile = [iASL decompileAML:[iASL fetchTable:table] name:table tableset:kSystemTableset refs:nil];
        if (!ModalError(decompile.error))
            [(document = [self newDocument:decompile.string displayName:nil tableName:table tableset:kSystemTableset display:false]) setFileURL:url];
    }
    else
        document = [super makeDocumentWithContentsOfURL:url ofType:typeName error:outError];
    if (!document && outError)
        *outError = [NSError errorWithDomain:kMaciASLDomain code:kTablesetError userInfo:nil];
    return document;
}

-(BOOL)presentError:(NSError *)error {
    if (error.code == kTablesetError && [error.domain isEqualToString:kMaciASLDomain]) return false;
    else return [super presentError:error];
}

#pragma mark Tableset
-(Document *)openTablesetTableWithContentsOfURL:(NSURL *)url {
    NSDictionary *tables = [NSDictionary dictionaryWithContentsOfURL:url];
    NSString *prefix = [[tables objectForKey:@"Hostname"] stringByAppendingString:@" "];
    tables = [tables objectForKey:@"Tables"];
    _tableView.titleWithRepresentedFilename = [url path];
    _tableView.representedURL = url;
    NSMutableArray *tempNames = [NSMutableArray array], *tempTables = [NSMutableArray array];
    for (NSString *name in [tables.allKeys sortedArrayUsingSelector:@selector(localizedStandardCompare:)]) {
        NSString *type;
        if ([name hasPrefix:@"SSDT"])
            type = [[NSString alloc] initWithData:[[tables objectForKey:name] subdataWithRange:NSMakeRange(16, 8)] encoding:NSASCIIStringEncoding];
        [tempNames addObject:type ? [NSString stringWithFormat:@"%@ (%@)", name, type] : name];
        [tempTables addObject:name];
    }
    assignWithNotice(self, tableNames, [tempNames copy]);
    __block NSInteger modal;
    dispatch_sync(dispatch_get_main_queue(), ^{ modal = [NSApp runModalForWindow:self->_tableView]; });
    if (modal == NSRunAbortedResponse)
        return nil;
    else if (modal == NSRunStoppedResponse)
        tables = @{[tempTables objectAtIndex:_tableSelection]:[tables objectForKey:[tempTables objectAtIndex:_tableSelection]]};
    Document *document;
    modal = 0;
    for (NSString *table in tables) {
        modal++;
        iASLDecompilationResult *decompile = [iASL decompileAML:[tables objectForKey:table] name:table tableset:url refs:nil];
        if (!ModalError(decompile.error))
            document = [self newDocument:decompile.string displayName:[prefix stringByAppendingString:table] tableName:table tableset:url display:modal != tables.count];
    }
    return document;
}

-(Document *)openIOJonesTableWithContentsOfURL:(NSURL *)url {
    NSDictionary *tables = [NSDictionary dictionaryWithContentsOfURL:url];
    NSString *hostname = [(NSDictionary *)[tables objectForKey:@"system"] objectForKey:@"systemName"];
    for (NSDictionary *object in [tables objectForKey:@"objects"])
        if ([[object objectForKey:@"class"] isEqualToString:@"AppleACPIPlatformExpert"]) {
            tables = [(NSDictionary *)[object objectForKey:@"properties"] objectForKey:@"ACPI Tables"];
            break;
        }
    if (tables) {
        char *temp;
        mkdtemp(temp = strdup([[NSTemporaryDirectory() stringByAppendingPathComponent:@"tableset.XXXXXX"] fileSystemRepresentation]));
        url = [(__bridge_transfer NSURL *)CFURLCreateFromFileSystemRepresentation(kCFAllocatorDefault, (UInt8 *)temp, strlen(temp), true) URLByAppendingPathComponent:[NSString stringWithFormat:@"%@.acpi", hostname]];
        free(temp);
        if ([@{@"Hostname":hostname, @"Tables":tables} writeToURL:url atomically:true])
            return [self openTablesetTableWithContentsOfURL:url];
    }
    else
        ModalError([NSError errorWithDomain:kMaciASLDomain code:kTablesetError userInfo:@{NSLocalizedDescriptionKey:@"IOJones Error", NSLocalizedRecoverySuggestionErrorKey:[NSString stringWithFormat:@"Could not find /AppleACPIPlatformExpert/ACPI Tables"]}]);
    return nil;
}

-(IBAction)finishTableset:(id)sender {
    if ([[sender title] isEqualToString:@"Cancel"])
        [NSApp abortModal];
    else if ([[sender title] isEqualToString:@"Open Selected"])
        [NSApp stopModal];
    else
        [NSApp stopModalWithCode:NSModalResponseContinue];
    [_tableView orderOut:sender];
}

#pragma mark New Documents
-(Document *)newDocument:(NSString *)text displayName:(NSString *)displayName tableName:(NSString *)tableName tableset:(NSURL *)tableset display:(bool)display {
    NSError *err;
    Document *doc = [[Document alloc] initWithType:kUTTypeAML tableName:tableName tableset:tableset text:text error:&err];
    if (ModalError(err))
        return nil;
    doc.displayName = displayName;
    [self addDocument:doc];
    if (display) {
        [doc makeWindowControllers];
        [doc performSelectorOnMainThread:@selector(showWindows) withObject:nil waitUntilDone:true];
    }
    return doc;
}

-(Document *)newDocumentFromACPI:(NSString *)name saveFirst:(bool)save {
    NSData *aml;
    if ((aml = [iASL fetchTable:name])) {
        NSURL *file = [iASL wasInjected:name];
        if (save && !file) {
            NSSavePanel *panel = [NSSavePanel savePanel];
            panel.prompt = @"Presave";
            panel.nameFieldStringValue = name;
            panel.allowedFileTypes = @[kUTTypeAML];
            if ([panel runModal] == NSFileHandlingPanelOKButton && [aml writeToURL:panel.URL atomically:true])
                file = panel.URL;
        }
        if (file && [[NSData dataWithContentsOfURL:file] isEqualToData:aml])
            [self openDocumentWithContentsOfURL:file display:true completionHandler:^(NSDocument *document, BOOL documentWasAlreadyOpen, NSError *error) {}];
        else {
            iASLDecompilationResult *decompile = [iASL decompileAML:aml name:name tableset:kSystemTableset refs:nil];
            if (!ModalError(decompile.error))
                return [self newDocument:decompile.string displayName:[NSString stringWithFormat:!file?@"System %@":@"System %@", name] tableName:name tableset:kSystemTableset display:true];
        }
    }
    return nil;
}

@end


MaciASL/AppDelegate.m
//
//  AppDelegate.m
//  MaciASL
//
//  Created by PHPdev32 on 9/27/12.
//  Licensed under GPLv3, full text at http://www.gnu.org/licenses/gpl-3.0.txt
//

#import "AppDelegate.h"
#import "Source.h"
#import "Document.h"
#import "Colorize.h"
#import "iASL.h"
#import "SSDT.h"
#import "DocumentController.h"
#import <sys/sysctl.h>
#import <sys/types.h>

@interface LogEntry : NSObject

@property NSDate *timestamp;
@property NSString *entry;

@end

@implementation LogEntry

-(instancetype)initWithEntry:(NSString *)entry {
    self = [super init];
    if (self) {
        _timestamp = [NSDate date];
        _entry = entry;
    }
    return self;
}

@end

@implementation AppDelegate {
    @private
    NSMutableArray *_log;
    __unsafe_unretained IBOutlet NSView *_general, __unsafe_unretained *_iasl, __unsafe_unretained *_sources;
    __unsafe_unretained IBOutlet NSWindow *_logView, __unsafe_unretained *_summaryView;
    __unsafe_unretained IBOutlet NSTableView *_sourceView;
    __unsafe_unretained IBOutlet NSArrayController *_sourceController;
}

#pragma mark Application Delegate
-(void)awakeFromNib {
    _log = [NSMutableArray array];
    [NSUserDefaults.standardUserDefaults registerDefaults:
    @{
      @"theme": @"Light",
      @"dsdt": @(YES),
      @"suggest": @(NO),
      @"acpi": @62,
      @"context": @(NO),
      @"isolation": @(NO),
      @"colorize": @(YES),
      @"remarks": @(NO),
      @"optimizations": @(NO),
      @"werror": @(NO),
      @"autoloadtables": @(NO),
      @"extracomp": @"",
      @"extradecomp": @"",
      @"preference": @0,
      @"font": @{@"name": @"Menlo", @"size": @11},
      @"sources": @[
        
        @{@"name": @"Xc2333", @"url": @"http://raw.github.com/Xc2333/Laptop-DSDT-Patch/master"},
        @{@"name": @"VoodooI2C", @"url": @"http://raw.github.com/alexandred/VoodooI2C-Patches/master"},
       
    ]}];
    NSFontManager.sharedFontManager.target = self;
    NSDictionary *font = [NSUserDefaults.standardUserDefaults objectForKey:@"font"];
    [NSFontManager.sharedFontManager setSelectedFont:[NSFont fontWithName:[font objectForKey:@"name"] size:[[font objectForKey:@"size"] floatValue]] isMultiple:false];
    _logView.level = NSNormalWindowLevel;
    [iASL addObserver:self forKeyPath:@"compiler" options:0 context:NULL];
}

-(BOOL)applicationShouldOpenUntitledFile:(NSApplication *)sender {
    if (![NSUserDefaults.standardUserDefaults boolForKey:@"dsdt"])
        return true;
    [DocumentController.sharedDocumentController newDocumentFromACPI:@"DSDT" saveFirst:false];
    return false;
}


#pragma mark Logging
-(void)logEntry:(NSString *)entry {
    if (NSThread.isMainThread) {
        insertWithNotice(self, log, [[LogEntry alloc] initWithEntry:entry])
    }
    else
        [self performSelectorOnMainThread:_cmd withObject:entry waitUntilDone:false];
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    muteWithNotice(self, compiler,);
    for (Document *doc in [NSDocumentController.sharedDocumentController documents])
        if (!doc.isDocumentEdited)
            [doc revertToContentsOfURL:doc.fileURL ofType:doc.fileType error:NULL];
}

#pragma mark Actions
-(IBAction)copy:(id)sender {
    NSResponder *obj = [[NSApp keyWindow] firstResponder];
    if (obj.class == NSTableView.class || obj.class == FSTableView.class || obj.class == NSOutlineView.class) {
        if (![(NSTableView *)obj numberOfSelectedRows]) return;
        bool viewBased = ([(NSTableView *)obj rowViewAtRow:[(NSTableView *)obj selectedRow] makeIfNecessary:false]);
        NSMutableArray *rows = [NSMutableArray array];
        [[(NSTableView *)obj selectedRowIndexes] enumerateIndexesUsingBlock:^void(NSUInteger idx, BOOL *stop){
            NSUInteger i = 0, j = [(NSTableView *)obj numberOfColumns];
            NSMutableArray *row = [NSMutableArray array];
            if (viewBased) {
                NSText *view;
                while (i < j)
                    if ((view = [(NSTableView *)obj viewAtColumn:i++ row:idx makeIfNecessary:false]) && [view isKindOfClass:NSText.class])
                        [row addObject:[view.string stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet]];
            }
            else {
                NSCell *cell;
                while (i < j)
                    if ((cell = [(NSTableView *)obj preparedCellAtColumn:i++ row:idx]) && [cell isKindOfClass:NSTextFieldCell.class])
                        [row addObject:[cell.stringValue stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet]];
            }
            [row removeObject:@""];
            [rows addObject:[row componentsJoinedByString:@", "]];
        }];
        [NSPasteboard.generalPasteboard clearContents];
        [NSPasteboard.generalPasteboard writeObjects:@[[rows componentsJoinedByString:@"\n"]]];
    }
}

-(IBAction)swapPreference:(id)sender {
    [self viewPreference:sender];
}

-(IBAction)documentFromACPI:(id)sender {
    [DocumentController.sharedDocumentController newDocumentFromACPI:[sender title] saveFirst:NSEvent.modifierFlags&NSAlternateKeyMask];
}

-(IBAction)showLog:(id)sender {
    [_logView makeKeyAndOrderFront:sender];
}

-(IBAction)showSummary:(id)sender {
    [_summaryView makeKeyAndOrderFront:sender];
}

-(IBAction)update:(id)sender {
    [sender setEnabled:false];
    muteWithNotice(self, update, _update = [NSProgress progressWithTotalUnitCount:3]);
    dispatch_group_t g = dispatch_group_create();
    for (NSNumber *iasl in @[@4, @5, @51, @6, @62]) {
        [_update becomeCurrentWithPendingUnitCount:1];
        dispatch_group_enter(g);
        NSURL *src = [NSURL URLWithString:[NSString stringWithFormat:@"https://github.com/acidanthera/MaciASL/raw/master/Dist/iasl%ld", iasl.unsignedIntegerValue]];
        NSURL *dst = [[[NSBundle.mainBundle executableURL] URLByDeletingLastPathComponent] URLByAppendingPathComponent:[NSString stringWithFormat:@"iasl%ld", iasl.unsignedIntegerValue]];
        [URLTask conditionalGet:src toURL:dst perform:^(bool success){
            muteWithNotice(self->_update, fractionCompleted,);
            if (success && [NSUserDefaults.standardUserDefaults integerForKey:@"acpi"] == iasl.unsignedIntegerValue)
                [iASL observeValueForKeyPath:nil ofObject:nil change:nil context:nil];
            dispatch_group_leave(g);
        }];
        [_update resignCurrent];
    }
    dispatch_group_notify(g, dispatch_get_main_queue(), ^{
        muteWithNotice(self, update, self->_update = nil);
        [sender setEnabled:true];
    });
}

-(IBAction)newSource:(id)sender {
    [_sourceController insertObject:_sourceController.newObject atArrangedObjectIndex:0];
    [_sourceView selectRowIndexes:[NSIndexSet indexSetWithIndex:0] byExtendingSelection:false];
    [_sourceView editColumn:0 row:0 withEvent:nil select:true];
}

-(IBAction)exportTableset:(id)sender {
    NSSavePanel *save = [NSSavePanel savePanel];
    save.prompt = @"Export Tableset";
    save.nameFieldStringValue = NSHost.currentHost.localizedName;
    save.allowedFileTypes = @[kUTTypeTableset];
    if ([save runModal] == NSFileHandlingPanelOKButton) {
        NSError *err;
        NSData *data = [NSPropertyListSerialization dataWithPropertyList:@{@"Hostname":NSHost.currentHost.localizedName, @"Tables":iASL.tableset} format:NSPropertyListBinaryFormat_v1_0 options:0 error:&err];
        if (ModalError(err)) return;
        [data writeToURL:save.URL atomically:true];
    }
}

#pragma mark Readonly Properties
-(NSString *)compiler {
    return iASL.compiler;
}

-(NSArray *)deviceProperties {
    return iASL.deviceProperties;
}

-(NSArray *)log {
    return [_log copy];
}

-(NSArray *)logAtIndexes:(NSIndexSet *)indexes {
    return [_log objectsAtIndexes:indexes];
}

-(id)objectInLogAtIndex:(NSUInteger)index {
    return [_log objectAtIndex:index];
}

-(NSArray *)themes {
    return ColorTheme.allThemes.allKeys;
}

#pragma mark Functions
-(void)viewPreference:(id)sender {
    NSWindow *preferences = [NSApp keyWindow];
    NSUInteger index;
    if (!sender) {
        index = [NSUserDefaults.standardUserDefaults integerForKey:@"preference"];
        sender = [preferences.toolbar.items objectAtIndex:index];
    }
    else {
        index = [preferences.toolbar.items indexOfObject:sender];
        if (index != NSNotFound)
            [NSUserDefaults.standardUserDefaults setInteger:index forKey:@"preference"];
    }
    NSSize newSize;
    switch (index) {
        case 0:
            newSize = _general.frame.size;
            preferences.contentView = _general;
            preferences.contentSize = newSize;
            break;
        case 1:
            newSize = _iasl.frame.size;
            preferences.contentView = _iasl;
            preferences.contentSize = newSize;
            break;
        case 2:
            newSize = _sources.frame.size;
            preferences.contentView = _sources;
            preferences.contentSize = newSize;
            break;
        default:
            return;
    }
    preferences.toolbar.selectedItemIdentifier = [sender itemIdentifier];
}

-(void)changeFont:(id)sender {
    NSFontManager *manager = NSFontManager.sharedFontManager;
    NSFont *font = [manager convertFont:[manager selectedFont]];
    [NSUserDefaults.standardUserDefaults setObject:@{@"name":font.displayName, @"size":@(font.pointSize)} forKey:@"font"];
    muteWithNotice(manager, selectedFont, [manager setSelectedFont:font isMultiple:false])
}

#pragma mark NSTableViewDelegate
-(void)tableViewSelectionDidChange:(NSNotification *)notification {
    [[NSDocumentController.sharedDocumentController documentForWindow:[NSApp mainWindow]] tableViewSelectionDidChange:notification];
}

-(CGFloat)outlineView:(NSOutlineView *)outlineView heightOfRowByItem:(id)item {
    NSInteger rows = outlineView.tableColumns.count, row = [outlineView rowForItem:item];
    CGFloat height = outlineView.rowHeight;
    while (rows-- > 0)
        height = MAX([[outlineView preparedCellAtColumn:rows row:row] cellSizeForBounds:NSMakeRect(0, 0, [[outlineView.tableColumns objectAtIndex:rows] width], CGFLOAT_MAX)].height,height);
    return height;
}

#pragma mark NSWindowDelegate
-(void)windowDidBecomeKey:(NSNotification *)notification {
    if ([notification.object isMemberOfClass:NSWindow.class])
        [self viewPreference:nil];
}
@end

@implementation FSTableView

-(BOOL)acceptsFirstMouse:(NSEvent *)theEvent {
    return true;
}

@end

@implementation FSPanel

-(BOOL)becomesKeyOnlyIfNeeded {
    return true;
}

@end

@interface NSTextFinder ()

+(NSTextFinder *)_globalTextFinder;

@end

@implementation FSTextView

-(void)scrollRangeToVisible:(NSRange)range {
    [super scrollRangeToVisible:range];
    id delegate = self.delegate;
    if (!NSEqualRanges(range, self.selectedRange)
        && NSTextFinder._globalTextFinder.client == (id)self
        && delegate && [delegate conformsToProtocol:@protocol(NSTextFinderIndication)])
        [(id<NSTextFinderIndication>)delegate textViewDidShowFindIndicator:[NSNotification notificationWithName:@"NSTextViewDidShowFindIndicatorNotification" object:self userInfo:@{@"NSFindIndicatorRange":[NSValue valueWithRange:range]}]];
}

@end

@implementation FSRulerView
static NSDictionary *style;

+(void)load {
    NSMutableParagraphStyle *temp = [NSMutableParagraphStyle new];
    temp.alignment = NSRightTextAlignment;
    NSFont *font = nil;
#ifdef __MAC_10_11
    if ([NSFont respondsToSelector:@selector(monospacedDigitSystemFontOfSize:weight:)])
        font = [NSFont monospacedDigitSystemFontOfSize:NSFont.smallSystemFontSize weight:NSFontWeightRegular];
    else
        font = [NSFont systemFontOfSize:NSFont.smallSystemFontSize];
#else
    font = [NSFont systemFontOfSize:NSFont.smallSystemFontSize];
#endif
    style = @{NSFontAttributeName:font, NSParagraphStyleAttributeName:[temp copy]};
}

-(instancetype)init {
    self = [super init];
    if (self)
        super.reservedThicknessForMarkers = 0;
    return self;
}

-(void)drawHashMarksAndLabelsInRect:(NSRect)rect {
    NSScrollView *scrollView = self.scrollView;
    if (!scrollView) return;
    NSInteger height = (NSInteger)[[(NSTextView *)scrollView.documentView layoutManager] defaultLineHeightForFont:NSFontManager.sharedFontManager.selectedFont], start = (NSInteger)floor((scrollView.documentVisibleRect.origin.y + rect.origin.y) / height) + 1, stop = 1 + start + (NSInteger)ceil(rect.size.height / height);
    if (self.ruleThickness < MAX(16,((NSInteger)log10(stop)+1)*8)) {
        self.ruleThickness = ((NSInteger)log10(stop)+1)*8;
        return;
    }
    rect.size.width -= 2;
    rect.origin.y -= (NSInteger)(scrollView.documentVisibleRect.origin.y+rect.origin.y) % height - (height-(NSFont.smallSystemFontSize+2))/2 + 1;
    rect.size.height = height;
    while (start < stop) {
        [[NSString stringWithFormat:@"%ld", start++] drawWithRect:rect options:NSStringDrawingUsesLineFragmentOrigin attributes:style];
        rect.origin.y += height;
    }
}

@end

    
